spec/machine/ARM/MachineOps.thy

```diff
@@ -184,53 +266,81 @@ lemma replicate_no_overflow:
   by (metis (mono_tags, hide_lams) le_unat_uoi of_nat_mult word_of_nat word_unat.Rep_inverse)
 
 lemma kernelWCET_ticks_pos2: "0 < 2 * kernelWCET_ticks"
-  apply (simp add: kernelWCET_ticks_def word_less_nat_alt)
-  apply (subgoal_tac "unat (2 * us_to_ticks kernelWCET_us) = 2 * unat (us_to_ticks kernelWCET_us)")
-   using ARM.kernelWCET_us_pos2 ARM.us_to_ticks_nonzero unat_gt_0 apply force
-  apply (subgoal_tac "2 * unat (us_to_ticks kernelWCET_us) \<le> unat (max_word :: ticks)")
-   using replicate_no_overflow apply fastforce
-  using kernelWCET_ticks_no_overflow by force
+  apply (simp add: kernelWCET_ticks_def)
+  using MIN_BUDGET_pos' by simp
 
 lemma getCurrentTime_buffer_nonzero':
   "0 < getCurrentTime_buffer"
-  apply (insert getCurrentTime_buffer_nonzero)
+  apply (insert getCurrentTime_buffer_pos)
   apply (clarsimp simp: kernelWCET_ticks_def MAX_PERIOD_def)
   done
 
-lemma getCurrentTime_buffer_no_overflow':
-  "unat kernelWCET_ticks + 5 * unat MAX_PERIOD = unat (kernelWCET_ticks + 5 * MAX_PERIOD)"
-  apply (prop_tac "5 * unat (us_to_ticks MAX_PERIOD_US) = unat (5 * us_to_ticks MAX_PERIOD_US)")
-   apply (prop_tac "5 * unat (us_to_ticks MAX_PERIOD_US) \<le> unat (max_word :: 64 word)")
-    using getCurrentTime_buffer_no_overflow apply linarith
-   apply (fastforce dest: replicate_no_overflow[where a="us_to_ticks MAX_PERIOD_US" and n=5])
-  apply (insert getCurrentTime_buffer_no_overflow)
-  apply (clarsimp simp: kernelWCET_ticks_def MAX_PERIOD_def)
-  apply (subst unat_add_lem')
-   apply (clarsimp simp: max_word_def)
-  apply simp
-  done
+lemma us_to_ticks_helper:
+  "unat a * unat ticks_per_timer_unit \<le> unat max_time
+   \<Longrightarrow> unat (us_to_ticks a) \<le> unat a * unat ticks_per_timer_unit"
+  apply (clarsimp simp: us_to_ticks_def)
+  by (subst unat_div | subst unat_mult_lem' | simp)+
+
+lemma le_mono:
+  "\<lbrakk>a \<le> a'; b \<le> b'; a' + b' \<le> c\<rbrakk> \<Longrightarrow> a + b \<le> (c :: nat)"
+  by fastforce
+
+lemma getCurrentTime_buffer_no_overflow:
+  "unat kernelWCET_ticks + 5 * unat MAX_PERIOD < unat max_time"
```

Comment:
This is a bit of a complicated area. The constant `5` is currently used in both `getCurrentTime` and `MAX_RELEASE_TIME`, but it is coincidental that they both do use `5`. It is likely that we'll want to change these in the future, but also likely they'll need to be different values at some point. 

We could introduce an alias for this constant factor and use it in the axiomatisation and the definition of `getCurrentTime_buffer`, but its usefullness might be limited; we currently have a lot of lemmas that use `current_time_bounded n` for some `n` less than 5. We may want to (try to) make everything the same constant in lemma statements (or just remove the constant as a parameter to `current_time_bounded`), but I'm not sure whether this will be possible, easy, or all that helpful

Reply:
Thanks for the comment, and, as we discussed in the meeting, let’s leave these as is for now.